connector/src/main/java/org/geysermc/connector/entity/ThrowableEntity.java

```diff
@@ -26,11 +26,65 @@
 package org.geysermc.connector.entity;
 
 import com.nukkitx.math.vector.Vector3f;
+import com.nukkitx.protocol.bedrock.data.LevelEventType;
+import com.nukkitx.protocol.bedrock.data.entity.EntityFlag;
+import com.nukkitx.protocol.bedrock.packet.LevelEventPacket;
 import org.geysermc.connector.entity.type.EntityType;
+import org.geysermc.connector.network.session.GeyserSession;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
 
 public class ThrowableEntity extends Entity {
 
+    private Vector3f lastPosition;
+    private ScheduledFuture<?> positionUpdater;
+
     public ThrowableEntity(long entityId, long geyserId, EntityType entityType, Vector3f position, Vector3f motion, Vector3f rotation) {
         super(entityId, geyserId, entityType, position, motion, rotation);
+        this.lastPosition = position;
+    }
+
+    @Override
+    public void spawnEntity(GeyserSession session) {
+        super.spawnEntity(session);
+        positionUpdater = session.getConnector().getGeneralThreadPool().scheduleAtFixedRate(() -> {
+            super.moveRelative(session, motion.getX(), motion.getY(), motion.getZ(), getRotation(), isOnGround());
+
+            if (getMetadata().getFlags().getFlag(EntityFlag.HAS_GRAVITY)) {
+                float gravity = 0.03f; // Snowball, Egg, and Ender Pearl,
+                if (getEntityType() == EntityType.THROWN_POTION || getEntityType() == EntityType.LINGERING_POTION) {
+                    gravity = 0.05f;
+                } else if (getEntityType() == EntityType.THROWN_EXP_BOTTLE) {
+                    gravity = 0.07f;
+                }
+                setMotion(getMotion().down(gravity));
+            }
+        }, 0, 50, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public boolean despawnEntity(GeyserSession session) {
+        positionUpdater.cancel(true);
+        if (getEntityType() == EntityType.THROWN_ENDERPEARL) {
+            LevelEventPacket particlePacket = new LevelEventPacket();
+            particlePacket.setType(LevelEventType.PARTICLE_TELEPORT);
+            particlePacket.setPosition(getPosition());
+            session.sendUpstreamPacket(particlePacket);
+        }
+        return super.despawnEntity(session);
+    }
+
+    @Override
+    public void moveRelative(GeyserSession session, double relX, double relY, double relZ, Vector3f rotation, boolean isOnGround) {
+        setPosition(lastPosition);
```

Comment:
Not sure why you have to do this? `moveRelative` sets the position, just as `this.position = ...`